File: pubnub-java/src/integrationTest/java/com/pubnub/api/integration/StorageAndPlaybackIntegrationTests.java
package com.pubnub.api.integration;

import com.pubnub.api.integration.util.BaseIntegrationTest;
import com.pubnub.api.integration.util.RandomGenerator;
import org.junit.Test;

import java.util.Collections;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class StorageAndPlaybackIntegrationTests extends BaseIntegrationTest {

    @Override
    protected void onBefore() {

    }

    @Override
    protected void onAfter() {

    }

    @Test
    public void testHistoryMessages() {
        final AtomicBoolean success = new AtomicBoolean();
        final String messageText = RandomGenerator.newValue(10);
        final String channel = RandomGenerator.newValue(10);

        subscribeToChannel(pubNub, channel);
        publishMessage(pubNub, channel, messageText);
        pause(1);

        pubNub.history()
                .channel(channel)
                .async((result, status) -> {
                    assertNotNull(result);
                    final String message = result.getMessages().get(0).getEntry().toString();
                    assertFalse(status.isError());
                    assertTrue(message.contains(pubNub.getConfiguration().getUserId().getValue()));
                    assertTrue(message.contains(messageText));
                    success.set(true);
                });

        success.set(true);
    }

    @Test
    public void testHistoryMessagesWithTimeToken() {
        final AtomicBoolean success = new AtomicBoolean();
        final String channel = RandomGenerator.newValue(10);

        pubNub.subscribe()
                .channels(Collections.singletonList(channel))
                .execute();

        publishMessages(channel, 3);
        pause(1);

        pubNub.history()
                .channel(channel)
                .includeTimetoken(true)
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assertNotNull(result);
                    int timeTokenCounter = 0;
                    for (int i = 0; i < result.getMessages().size(); i++) {
                        if (result.getMessages().get(i).getTimetoken() != null) {
                            timeTokenCounter++;
                        }
                    }

                    assertEquals(3, timeTokenCounter);
                    success.set(true);
                });

        success.set(true);
    }

    @Test
    public void testLoadingHistoryMessagesWithLimit() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = RandomGenerator.newValue(10);

        subscribeToChannel(pubNub, expectedChannel);
        publishMessages(expectedChannel, 20);
        pause(1);

        pubNub.history()
                .channel(expectedChannel)
                .count(10)
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assertNotNull(result);
                    final int numberOfMessages = result.getMessages().size();
                    assertEquals(10, numberOfMessages);
                    success.set(true);
                });

        success.set(true);
    }

    @Test
    public void testLoadingHistoryWithSpecificTimeInterval() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = RandomGenerator.newValue(10);

        subscribeToChannel(pubNub, expectedChannel);

        final Long before = System.currentTimeMillis() * 10_000;
        pause(5);
        publishMessages(expectedChannel, 3);
        pause(5);
        final Long now = System.currentTimeMillis() * 10_000;

        pubNub.history()
                .channel(expectedChannel)
                .includeTimetoken(true)
                .start(now)
                .end(before)
                .count(10)
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assertNotNull(result);
                    assertEquals(3, result.getMessages().size());
                    success.set(true);
                });
        success.set(true);
    }

    @Test
    public void testReverseHistoryPaging() {
        final AtomicBoolean success = new AtomicBoolean();
        final String channel = RandomGenerator.newValue(10);
        final String message_1 = RandomGenerator.newValue(20);
        final String message_2 = RandomGenerator.newValue(20);

        subscribeToChannel(pubNub, channel);
        publishMessage(pubNub, channel, message_1);
        publishMessage(pubNub, channel, message_2);
        pause(1);

        pubNub.history()
                .channel(channel)
                .count(10)
                .reverse(true)
                .async((result, status) -> {
                    assertFalse(status.isError());
                    assertNotNull(result);
                    if (!result.getMessages().isEmpty()) {
                        final String message = result.getMessages().get(0).getEntry().toString();
                        assertTrue(message.contains(message_1));
                    } else {
                        fail("Messages are empty");
                    }
                    success.set(true);
                });

        success.set(true);
    }

    private void publishMessages(String channel, int counter) {
        for (int i = 0; i < counter; i++) {
            publishMessage(pubNub, channel, RandomGenerator.newValue(10));
        }
    }
}

File: core/src/integrationTest/kotlin/com/pubnub/api/integration/StorageAndPlaybackIntegrationTests.kt
package com.pubnub.api.integration

import com.pubnub.api.CommonUtils.randomChannel
import com.pubnub.api.CommonUtils.randomValue
import com.pubnub.api.asyncRetry
import com.pubnub.api.await
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Test

class StorageAndPlaybackIntegrationTests : BaseIntegrationTest() {

    @Test
    fun testHistoryMessages() {
        val expectedMessage = randomValue()
        val expectedChannel = randomChannel()

        pubnub.publish(
            channel = expectedChannel,
            message = expectedMessage
        ).await { _, _ -> }

        pubnub.history(
            channel = expectedChannel
        ).asyncRetry { result, status ->
            assertFalse(status.error)
            assertEquals(expectedMessage, result!!.messages[0].entry.asString)
        }
    }

    @Test
    fun testHistoryMessagesWithTimeToken() {
        val expectedChannel = randomChannel()

        repeat(3) {
            pubnub.publish(
                channel = expectedChannel,
                message = randomValue()
            ).sync()!!
        }

        pubnub.history(
            channel = expectedChannel,
            includeTimetoken = true
        ).asyncRetry { result, status ->
            assertFalse(status.error)
            result!!.messages.forEach {
                assertNotNull(it.timetoken)
            }
        }
    }

    @Test
    fun testLoadingHistoryMessagesWithLimit() {
        val expectedChannel = randomChannel()

        repeat(20) {
            pubnub.publish(
                channel = expectedChannel,
                message = randomValue()
            ).sync()!!
        }

        pubnub.history(
            channel = expectedChannel,
            count = 10
        ).asyncRetry { result, status ->
            assertFalse(status.error)
            assertEquals(10, result!!.messages.size)
        }
    }

    @Test
    fun testLoadingHistoryWithSpecificTimeInterval() {
        val expectedChannel = randomChannel()

        val before = System.currentTimeMillis() * 10000
        wait(5)
        repeat(3) {
            pubnub.publish(
                channel = expectedChannel,
                message = randomValue()
            ).sync()!!
        }
        wait(5)
        val now = System.currentTimeMillis() * 10000

        pubnub.history(
            channel = expectedChannel,
            includeTimetoken = true,
            start = now,
            end = before,
            count = 10
        ).await { result, status ->
            assertFalse(status.error)
            assertEquals(3, result!!.messages.size)
        }
    }

    @Test
    fun testReverseHistoryPaging() {
        val expectedChannel: String = randomValue()
        val message1: String = randomValue(20)
        val message2: String = randomValue(20)

        pubnub.publish(
            channel = expectedChannel,
            message = message1
        ).sync()!!

        pubnub.publish(
            channel = expectedChannel,
            message = message2
        ).sync()!!

        pubnub.history(
            channel = expectedChannel,
            count = 10,
            reverse = true
        ).asyncRetry { result, status ->
            assertFalse(status.error)
            assertEquals(message1, result!!.messages[0].entry.asString)
            assertEquals(message2, result.messages[1].entry.asString)
        }
    }
}
